Proxy/HTTP/HighlightResponsesWithDeveloperNotes

```diff
@@ -0,0 +1,61 @@
+/**
+ * Bambda Script to Highlight Responses with Developer Notes
+ * This script identifies and highlights HTTP responses containing developer notes in HTML, JavaScript, or other files.
+ * It differentiates the types of files and highlights them accordingly: green for HTML, yellow for JavaScript, and blue for other types.
+ * Author: Tur24Tur
+ * GitHub: @BugBountyzip (https://github.com/BugBountyzip)
+ **/
+
+boolean manualColorHighlightEnabled = true;
+Set<String> ignoredExtensions = Set.of("mp4", "mp3", "png", "gif", "jpg", "jpeg", "css", "pdf");
+
+if (!requestResponse.hasResponse()) {
+    return false;
+}
+
+String requestUrl = requestResponse.request().url().toString();
+String fileExtension = requestUrl.substring(requestUrl.lastIndexOf('.') + 1).toLowerCase();
+
+if (ignoredExtensions.contains(fileExtension)) {
+    return false;
+}
+
+String contentType = requestResponse.response().headerValue("Content-Type");
+boolean isHtml = contentType != null && contentType.toLowerCase().contains("text/html");
+boolean isJavaScript = contentType != null && contentType.toLowerCase().contains("application/javascript");
+boolean foundDeveloperNotes = false;
+StringBuilder notesBuilder = new StringBuilder();
+HighlightColor highlightColor = HighlightColor.BLUE; // Default color
+
+if (isHtml || fileExtension.equals("html") || fileExtension.equals("htm")) {
+    highlightColor = HighlightColor.GREEN;
+} else if (isJavaScript || fileExtension.equals("js")) {
+    highlightColor = HighlightColor.YELLOW;
+}
+
+String responseBody = requestResponse.response().bodyToString();
+String[] commentPatterns = {"<!--(.*?)-->", "/[*][*](.*?)[*][*]/"};
+
+for (String pattern : commentPatterns) {
+    Pattern regexPattern = Pattern.compile(pattern, Pattern.DOTALL);
+    Matcher matcher = regexPattern.matcher(responseBody);
+
+    while (matcher.find()) {
+        foundDeveloperNotes = true;
+        if (manualColorHighlightEnabled) {
+            if (notesBuilder.length() > 0) {
+                notesBuilder.append("; ");
+            }
+            notesBuilder.append("Developer note found: ").append(matcher.group());
```

Comment:
When testing, I noticed a few false positives where large JS bundle files which contained logic within them related to comments included literal `/*` etc characters within strings. This lead to notes being added which were thousands of characters long. Have you had a look at any way to avoid this? Alternatively, it's probably worth capping the length of each developer note to around 250 characters or so to reduce the impact of false positives.

Reply:
Dear @ps-porpoise ,


For your kind review


```
boolean manualColorHighlightEnabled = true;

// Ensure there is a response and it is not null
if (!requestResponse.hasResponse() || requestResponse.response() == null) {
    return false;
}

// Use mimeType() for content type detection
MimeType responseType = requestResponse.response().mimeType();
boolean isHtml = responseType == MimeType.HTML;
boolean isJavaScript = responseType == MimeType.SCRIPT;
boolean foundDeveloperNotes = false;
StringBuilder notesBuilder = new StringBuilder();
HighlightColor highlightColor = HighlightColor.BLUE; // Default color for other types

if (isHtml) {
    highlightColor = HighlightColor.GREEN;
} else if (isJavaScript) {
    highlightColor = HighlightColor.YELLOW;
}

String responseBody = requestResponse.response().bodyToString();
String[] commentPatterns = {"<!--(.*?)-->", "/[*][*](.*?)[*][*]/"};

for (String pattern : commentPatterns) {
    Pattern regexPattern = Pattern.compile(pattern, Pattern.DOTALL);
    Matcher matcher = regexPattern.matcher(responseBody);

    while (matcher.find()) {
        foundDeveloperNotes = true;
        if (manualColorHighlightEnabled) {
            String note = matcher.group();
            // Limit the note length to 250 characters to avoid excessively long notes
            if (note.length() > 250) {
                note = note.substring(0, 250) + "...";
            }

            if (notesBuilder.length() > 0) {
                notesBuilder.append("; ");
            }
            notesBuilder.append("Developer note found: ").append(note);
        }
    }
}

if (foundDeveloperNotes) {
    requestResponse.annotations().setHighlightColor(highlightColor);
    if (manualColorHighlightEnabled && notesBuilder.length() > 0) {
        requestResponse.annotations().setNotes(notesBuilder.toString());
    }
}

return foundDeveloperNotes;



Changes Made

    Content Type Detection: Replaced manual content type checks with response().mimeType() for more streamlined and accurate detection.
    Simplified Highlighting Logic: Removed file extension checks, relying solely on MIME type.
    Limiting Note Length: Added a cap to the length of notes added for each found comment to prevent excessively long notes, which can be caused by false positives in large files.

These modifications should address the feedback points, making the script more efficient and reducing the likelihood of encountering issues with large JavaScript files.

```

Reply:
Another update

Content Type Filtering: The script now only processes responses with MIME types MimeType.HTML or MimeType.SCRIPT.

```
/**
 * Bambda Script to Highlight Responses with Developer Notes
 * This script identifies and highlights HTTP responses containing developer notes in HTML and JavaScript files.
 * It highlights HTML responses in green and JavaScript responses in yellow.
 * @author Tur24Tur / BugBountyzip (https://github.com/BugBountyzip)
 **/

boolean manualColorHighlightEnabled = true;

// Ensure there is a response and it is not null
if (!requestResponse.hasResponse() || requestResponse.response() == null) {
    return false;
}

// Use mimeType() for content type detection
MimeType responseType = requestResponse.response().mimeType();
boolean isHtml = responseType == MimeType.HTML;
boolean isJavaScript = responseType == MimeType.SCRIPT;

// Process only HTML and JavaScript responses
if (!isHtml && !isJavaScript) {
    return false;
}

boolean foundDeveloperNotes = false;
StringBuilder notesBuilder = new StringBuilder();
HighlightColor highlightColor = isHtml ? HighlightColor.GREEN : HighlightColor.YELLOW;

String responseBody = requestResponse.response().bodyToString();
String[] commentPatterns = isHtml ? new String[]{"<!--(?!\\[if).*?(?<!\\])-->"} : new String[]{"/[*][*](.*?)[*][*]/"};


for (String pattern : commentPatterns) {
    Pattern regexPattern = Pattern.compile(pattern, Pattern.DOTALL);
    Matcher matcher = regexPattern.matcher(responseBody);

    while (matcher.find()) {
        foundDeveloperNotes = true;
        if (manualColorHighlightEnabled) {
            String note = matcher.group();
            // Limit the note length to 250 characters
            if (note.length() > 250) {
                note = note.substring(0, 250) + "...";
            }

            if (notesBuilder.length() > 0) {
                notesBuilder.append("; ");
            }
            notesBuilder.append("Developer note found: ").append(note);
        }
    }
}

if (foundDeveloperNotes) {
    requestResponse.annotations().setHighlightColor(highlightColor);
    if (manualColorHighlightEnabled && notesBuilder.length() > 0) {
        requestResponse.annotations().setNotes(notesBuilder.toString());
    }
}

return foundDeveloperNotes;


```


i Also updated the regex pattern to be more selective , This pattern attempts to exclude HTML conditional comments like <!--[if IE 9]>...<![endif]--> by using negative lookaheads ((?!...)) and lookbehinds ((?<!...)).


String[] commentPatterns = isHtml ? new String[]{"<!--(?!\\[if).*?(?<!\\])-->"} : new String[]{"/[*][*](.*?)[*][*]/"};